Таратухин Сергей

taratuhin opened the pull request.

taratuhin committed:
Добавлен проект.

taratuhin committed:
Добавлена проверка что на кнопке найден объект Timer

Toxin65 commented on WidgetMovement/mainwindow.h:
```diff
@@ -0,0 +1,22 @@
+    ~MainWindow();
+
+private:
+    QTimer* m_timer = nullptr;
```
Это лишнее. Точно.

taratuhin replied:
Что именно лишнее (строки с 15 по 18 были в шаблоне)?

Toxin65 replied:
Хранение указателя на таймер не нужно.

Toxin65 replied:
Ну если он есть в шаблоне (я уже не помню), то пусть будет...

Toxin65 commented on WidgetMovement/mainwindow.cpp:
```diff
@@ -0,0 +1,85 @@
+                if (x > this->width() - pb->width())
+                {
+                    x = this->width() - pb->width();
+                }
```
Очень странная конструкция. Можно генерировать в нужном диапазоне, и брать "большой" генератор случайных чисел.
Это надо переделать. Однозначно.

Toxin65 commented on WidgetMovement/mainwindow.cpp:
```diff
@@ -0,0 +1,85 @@
+                }
+
+                pb->move(x, qrand() % 100);
+                pb->installEventFilter(this);
```
Что-то лишнее.

Toxin65 commented on WidgetMovement/mainwindow.cpp:
```diff
@@ -0,0 +1,85 @@
+                connect(pb, &QPushButton::clicked, [pb, tm] {
+                    tm->deleteLater();
+                    pb->deleteLater();
+                });
```
При правильно построеном владении это не нужно.

Toxin65 commented on WidgetMovement/mainwindow.cpp:
```diff
@@ -0,0 +1,85 @@
+                pb->show();
+
+                auto tm = new QTimer(pb);
+                tm->setInterval(qrand() % 100 + 50);
```
https://doc.qt.io/qt-6/qrandomgenerator.html#bounded-2

Toxin65 commented on WidgetMovement/mainwindow.cpp:
```diff
@@ -0,0 +1,85 @@
+    resize(300, 600);
+    m_timer = new QTimer();
+    m_timer->setInterval(1000);
+    qsrand(QDateTime::currentMSecsSinceEpoch() / 1000);
```
Это лишнее всё, есть глобальный генератор.

taratuhin replied:
qsrand(QDateTime::currentMSecsSinceEpoch() / 1000); убрал, но остальное осталось (размер экрана и создание таймера, не понимаю как можно таймер заменить глобальным генератором https://doc.qt.io/qt-6/qrandomgenerator.html#global)

taratuhin replied:
m_timer нужно (можно) заменить на QTimer::singleShot?

Toxin65 replied:
Нет, таймер нужен. Менять его на singleShot не надо. Комментарий был про одну строку.

Toxin65 commented on WidgetMovement/mainwindow.cpp:
```diff
@@ -0,0 +1,85 @@
+
+MainWindow::~MainWindow()
+{
+    m_timer->deleteLater();
```
Если есть владение -- это не нужно.

Toxin65 reviewed (changes requested):
Вы не понимаете как работают деревья владения.
И ещё куча замечаний.
Исправляйте. Вопросы -- пишите, можно в ЛС.

taratuhin committed:
Указатель на таймер m_timer больше не хранится в классе.
Переписал работу с случайными величинами .
Удалил лишний код.

Toxin65 commented on WidgetMovement/mainwindow.cpp:
```diff
@@ -0,0 +1,76 @@
+
+                connect(pb, &QPushButton::clicked, [pb] {
+                    pb->deleteLater();
+                });
```
Лямбда не нужна, обращение в `deleteLater` как в простой слот.

Toxin65 commented on WidgetMovement/mainwindow.cpp:
```diff
@@ -0,0 +1,76 @@
+                                this->setStyleSheet("background-color:red;");
+                                this->setWindowTitle("Game over!");
+                                tm->deleteLater();
+                                pb->deleteLater();
```
А почему два `deleteLater`?

Toxin65 commented on WidgetMovement/mainwindow.cpp:
```diff
@@ -0,0 +1,76 @@
+            [this, m_timer] {
+                auto pb = new QPushButton(this);
+                pb->resize(20, 20);
+                auto x = QRandomGenerator::global()->bounded(this->width() - pb->width());
```
const

Toxin65 commented on WidgetMovement/mainwindow.cpp:
```diff
@@ -0,0 +1,76 @@
+
+                connect(tm, &QTimer::timeout,
+                        [pb, tm, this] {
+                            pb->move(pb->pos() + QPoint(0, 1));
```
Не соответствует ТЗ. Нет реакции на наличие мыши над кнопкой.

taratuhin replied:
Какой пункт в ТЗ?

Toxin65 replied:
На самом деле сделано было через `MainWindow::eventFilter`, но когда я это читал, то ещё не понимал что так сделано.
Условие про то, что кнопка ускоряется, если над ней мышка.

Toxin65 commented on WidgetMovement/mainwindow.cpp:
```diff
@@ -0,0 +1,76 @@
+{
+}
+
+bool MainWindow::eventFilter(QObject *obj, QEvent *event)
```
Не соответствует условию ТЗ `- Запрещается создавать свои классы, наследовать, переопределять методы.`

Toxin65 reviewed (changes requested):
Нарушение условия ТЗ, некоторые ошибки.

taratuhin committed:
Вернул исходное состояние класса MainWindow (указатель на таймер в private, QTimer* m_timer)
Для случайной переменной отвечающей за появление кнопки по горизонтали (x) введён квалификатор const.
Переработан connect отвечающий за обработку нажатия кнопки.
Удалён запрос на удаление таймера (таймер уничтожается автоматически, так как является потомком кнопки).

taratuhin committed:
Переработан механизм ускорения, убрал механизма событий (eventFilter(…), QEvent::HoverEnter) и перешёл на метод underMouse()

Toxin65 commented on WidgetMovement/mainwindow.cpp:
```diff
@@ -0,0 +1,57 @@
+                                int interval = tm->interval();
+                                interval -= interval * 0.5;
+                                interval < 15 ? interval = 15 : interval;
+                                tm->setInterval(interval);
```
Таймер будет не в восторге, что его дёргают. Предлагаю удваивать перемещение, а не таймер передёргивать.
Если подумать что делается на уровне операционной системы -- скорее всего таймер пересоздаётся.

Toxin65 reviewed (approved):
В целом то, что нужно, за исключением комментария про таймер.
Ну и прикидывайте частоту обновления кадров на экране и периоды вашего таймера, если по взрослому думать.
